logger.go

```diff
@@ -0,0 +1,49 @@
+package main
+
+import (
+	"log"
+
+	"github.com/go-logr/logr"
+	"github.com/go-logr/zapr"
+	"github.com/urfave/cli/v2"
+	"go.uber.org/zap"
+	"go.uber.org/zap/zapcore"
+)
+
+const loggerMetadataKeyName = "logger"
+
+// AppLogger retrieves the application-wide logger instance from the cli.Context's Metadata.
+// This function will return nil if SetAppLogger was not called before this function is called.
+func AppLogger(c *cli.Context) logr.Logger {
+	return c.App.Metadata[loggerMetadataKeyName].(logr.Logger)
+}
+
+// SetAppLogger stores the application-wide logger instance to the cli.Context's Metadata,
+// so that it can later be retrieved by AppLogger.
+func SetAppLogger(c *cli.Context, logger logr.Logger) {
+	c.App.Metadata[loggerMetadataKeyName] = logger
```

Comment:
I think we should really save the logger in the native go context. So it's preserved when passing through `context.WIthCancel`, `context.WithValue`, etc.

Reply:
Turns out I can only create a context with initial values (there's no Setter). But at app creation time the logger isn't existing yet, only after starting and parsing the flags (in the before() func). Instead, I've opted to add the App instance to the context values, which can then be used to retrieve the logger additionally (and exposed their key names). So, if only the context.Context is available and not cli.Context, then it's still possible to retrieve the root logger. Is that acceptable?

Reply:
I'm torn. Everything in context should be thread-safe so this feels wrong. I'd have to think about a better solution though.

Reply:
logr for reference: https://pkg.go.dev/github.com/go-logr/logr#NewContext and https://pkg.go.dev/github.com/go-logr/logr#FromContextOrDiscard.